Suppressig generic

jtlap opened the pull request:
first pass just creating proper scalar versions

jfalcou commented on include/boost/simd/arch/common/scalar/function/load.hpp:
```diff
@@ -90,24 +90,6 @@ namespace boost { namespace simd { namespace ext
       return p.mask() ? boost::simd::load<target>(p.get(),o) : static_cast<target>(p.value());
     }
   };
-
```
why is it removed ?

jtlap replied:
seems to be a manip error

jfalcou commented on include/boost/simd/arch/common/scalar/function/genmask.hpp:
```diff
@@ -72,38 +72,38 @@ namespace boost { namespace simd { namespace ext
+//                           , bd::cpu_
+//                           , bd::scalar_< bs::logical_<A0> >
+//                           , bd::target_< bd::scalar_<bd::unspecified_<Target> > >
+//                           )
```
again why is this commented and not removed ?
Then, why do you need to touch a *scalar* version ?

jtlap replied:
corrected

jfalcou commented on include/boost/simd/arch/common/simd/function/divfloor.hpp:
```diff
@@ -33,18 +33,17 @@ namespace boost { namespace simd { namespace ext
-   };
-
+  {
+    BOOST_FORCEINLINE A0 operator()( bd::functor<bs::tag::floor_> const&
```
this commit is useless

jfalcou commented on include/boost/simd/arch/common/simd/function/mask2logical.hpp:
```diff
@@ -20,9 +20,8 @@ namespace boost { namespace simd { namespace ext
   namespace bs = boost::simd;
 
-  BOOST_DISPATCH_OVERLOAD_IF (  mask2logical_
+  BOOST_DISPATCH_OVERLOAD (  mask2logical_
```
why the revert ?

jfalcou commented on include/boost/simd/arch/common/generic/function/is_not_real.hpp:
```diff
@@ -27,7 +27,7 @@ namespace boost { namespace simd { namespace ext
   {
     using result = bs::as_logical_t<A0>;
-    BOOST_FORCEINLINE result operator() ( A0 const&) const BOOST_NOEXCEPT
+    BOOST_FORCEINLINE result operator() ( A0 ) const BOOST_NOEXCEPT
```
if thisis generic, you need the const&
read before comitting please.

jfalcou commented on include/boost/simd/arch/common/scalar/function/acos.hpp:
```diff
@@ -72,11 +64,23 @@ namespace boost { namespace simd { namespace ext
       //         2 values (0.00%) within 256.0 ULPs in range [9.999996424e-01, 9.999997020e-01]. Example: 9.999996424e-01 returns 8.456269861e-04 instead of 8.457279764e-04
       //         1 values (0.00%) within 512.0 ULPs in range [9.999999404e-01, 9.999999404e-01]. Example: 9.999999404e-01 returns 3.451863886e-04 instead of 3.452669771e-04
-      A0 z = Pio_2<A0>()-detail::invtrig_base<A0,tag::radian_tag,is_not_scalar_t<A0>>::asin(a0);
+      A0 z = Pio_2<A0>()-detail::invtrig_base<A0,tag::radian_tag,tag::not_simd_type>::asin(a0);
```
those tags are now maybe useless and can be repalced directly by std::true_type/false_type

jtlap replied:
ok, I will do it later as well in scalar as in simd when the generic will be put in simd

jfalcou replied:
OK perfect. The more unneeded code, the better.

jfalcou commented on include/boost/simd/arch/common/scalar/function/div.hpp:
```diff
@@ -52,12 +55,173 @@ namespace boost { namespace simd { namespace ext
+  #pragma warning(disable: 4723) // potential divide by 0
+#endif
+
+  namespace bd = boost::dispatch;
```
those liens are redundant I think

jtlap replied:
true

jfalcou commented on include/boost/simd/arch/common/scalar/function/abs_s.hpp:
```diff
@@ -0,0 +1,49 @@
+//==================================================================================================
+/**
+  Copyright 2016 NumScale SAS
+
```
you remember that those files should be included by their "parent" function (abs.hpp here) and not from the main include point right ?

jtlap replied:
yes

jfalcou commented on include/boost/simd/arch/common/scalar/function/all_reduce.hpp:
```diff
@@ -0,0 +1,136 @@
+  template<int Step> struct butterfly_perm
+  {
+    template<typename I, typename>
+    struct  apply : std::integral_constant<int,(I::value >= Step) ? I::value-Step : I::value+Step>
```
why does all_reduce , a pack only function, ends up with a scalar version ?

jtlap replied:
script side effect. I suppress it

jfalcou replied:
This is true for all the auto_* I guess so you can remove them in one scoop

jfalcou commented on include/boost/simd/arch/common/scalar/function/arg.hpp:
```diff
@@ -0,0 +1,75 @@
+     using hierarchy_tag = void; //dispatch::detail::hierarchy_tag;
+   };
+
+   const use_signbit_tag use_signbit_ = {};
```
what is this ?

jtlap replied:
arg has to versions one using the sign bit the other using the comparison to zero.
read the doc

jfalcou replied:
OK then it will be better if the use_signbit_ stuff could be put into the main include file with the documentation so we can document it.

jfalcou commented on include/boost/simd/arch/common/scalar/function/asecd.hpp:
```diff
@@ -0,0 +1,39 @@
+                            , bd::scalar_< bd::floating_<A0> >
+                            )
+  {
+    using result_t = A0;
```
useless using

jfalcou commented on include/boost/simd/arch/common/scalar/function/asecpi.hpp:
```diff
@@ -0,0 +1,39 @@
+                            , bd::scalar_< bd::floating_<A0> >
+                            )
+  {
+    using result_t = A0;
```
useless using

jfalcou commented on include/boost/simd/arch/common/scalar/function/asinpi.hpp:
```diff
@@ -0,0 +1,39 @@
+  {
+    BOOST_FORCEINLINE A0 operator() ( A0 a0) const BOOST_NOEXCEPT
+    {
+      return Invpi<A0>()*asin(a0);
```
for all those files, you deal with scalar, so asin is maybe the std one if the user had made a silly mistake of using namespace std. Prefix them with boost::simd::

jtlap replied:
done for a/cos a/sin a/tan abs. I will check the other later.

jfalcou commented on include/boost/simd/arch/common/scalar/function/autofold.hpp:
```diff
@@ -0,0 +1,96 @@
+  Copyright 2016 NumScale SAS
+
+  Distributed under the Boost Software License, Version 1.0.
+  (See accompanying file LICENSE.md or copy at http://boost.org/LICENSE_1_0.txt)
```
this file is also irrelevant

jfalcou commented on include/boost/simd/arch/common/scalar/function/automap.hpp:
```diff
@@ -0,0 +1,18 @@
+#ifndef BOOST_SIMD_ARCH_COMMON_SCALAR_FUNCTION_AUTOMAP_HPP_INCLUDED
+#define BOOST_SIMD_ARCH_COMMON_SCALAR_FUNCTION_AUTOMAP_HPP_INCLUDED
+
+#include <boost/simd/arch/common/generic/function/automap_opt.hpp>
```
why ???

jfalcou commented on include/boost/simd/arch/common/scalar/function/automap_decorator.hpp:
```diff
@@ -0,0 +1,180 @@
+  // -----------------------------------------------------------------------------------------------
+  // automap for functions with decorators
+  BOOST_DISPATCH_OVERLOAD_FALLBACK ( ( typename F
+                                     , typename Decorator, typename... Pn, typename... En
```
this file is also irrelevant

jfalcou commented on include/boost/simd/arch/common/scalar/function/automap_decorator_opt.hpp:
```diff
@@ -0,0 +1,189 @@
+**/
+//==================================================================================================
+#ifndef BOOST_SIMD_ARCH_COMMON_SCALAR_FUNCTION_AUTOMAP_DECORATOR_OPT_HPP_INCLUDED
+#define BOOST_SIMD_ARCH_COMMON_SCALAR_FUNCTION_AUTOMAP_DECORATOR_OPT_HPP_INCLUDED
```
this file is also irrelevant

jfalcou commented on include/boost/simd/arch/common/scalar/function/automap_opt.hpp:
```diff
@@ -0,0 +1,184 @@
+{
+  namespace bs = boost::simd;
+  namespace bd = boost::dispatch;
+  namespace br = brigand;
```
this file is also irrelevant

jfalcou commented on include/boost/simd/arch/common/scalar/function/automap_slice.hpp:
```diff
@@ -0,0 +1,113 @@
+#define BOOST_SIMD_ARCH_COMMON_SCALAR_FUNCTION_AUTOMAP_SLICE_HPP_INCLUDED
+
+#if BOOST_HW_SIMD_X86_OR_AMD >= BOOST_HW_SIMD_X86_AVX_VERSION
+#include <boost/simd/detail/overload.hpp>
```
this file is also irrelevant

jfalcou commented on include/boost/simd/arch/common/scalar/function/automap_value.hpp:
```diff
@@ -0,0 +1,181 @@
+#include <boost/simd/detail/dispatch/hierarchy/functions.hpp>
+#include <boost/simd/arch/common/tags.hpp>
+#include <boost/simd/detail/diagnostic.hpp>
+
```
this file is also irrelevant

jfalcou commented on include/boost/simd/arch/common/scalar/function/autoscan.hpp:
```diff
@@ -0,0 +1,156 @@
+namespace boost { namespace simd { namespace detail
+{
+  //------------------------------------------------------------------------------------------------
+  // This meta-permutation implements the scan pattern required for scan-trees.
```
this file is also irrelevant

jfalcou commented on include/boost/simd/arch/common/scalar/function/autosplat.hpp:
```diff
@@ -0,0 +1,321 @@
+                                   , bd::scalar_<bd::unspecified_<S0>>
+                                   , bs::pack_<bd::unspecified_<P0>, E0>
+                                   )
+  {
```
this file is also irrelevant

jfalcou commented on include/boost/simd/arch/common/scalar/function/bitwise_cast.hpp:
```diff
@@ -0,0 +1,46 @@
+  @file
+
+  Copyright 2015 NumScale SAS
+  Copyright 2015 J.T. Lapreste
```
also wrong headers, probably in all those new files

jfalcou commented on include/boost/simd/arch/common/scalar/function/comma.hpp:
```diff
@@ -0,0 +1,36 @@
+//==================================================================================================
+/*!
```
this function does nto exist anymore, why reintroducing this ?

jfalcou commented:
Stuff to fix:
 - you introduce a lot of files that make no sense
 - you need to prefix the call to our function because calling on scalars may default to std:: version
 - headers are all wrong
 - did you check your new files are ALL properly included in the PROPER files ? The ??_s.hpp variant may probably be merged with the ??.hpp and removed afterward to reduce include files
 - I am wary of your change on load and genmask, especially as it deals with function which are critical. Are you sure oyu need to remove them ?

jtlap commented:
to answer all these remarks

    you introduce a lot of files that make no sense

ok it was the script side effect as it took all generic files for which no scalar existed
I think it is now corrected

    you need to prefix the call to our function because calling on scalars may default to std:: version
    headers are all wrong

This is done now for a/cos a/sin a/tan and abs. I will look for other std functions
The headers are wrong because you modified then concurrently to my own modifcations
If you have a script, it seems better to run it once again before issuing the candidate-release.

    did you check your new files are ALL properly included in the PROPER files ? The ??_s.hpp variant may probably be merged with the ??.hpp and removed afterward to reduce include files

this have been done manually and the merging could be done (or not) later

    I am wary of your change on load and genmask, especially as it deals with function which are critical. Are you sure oyu need to remove them ?

There are no more changes. I reverted then for genmask and other files as load splat etc.
At the end of the process of splitting scalar/simd there will remain some generic files that will (will not need) special care

jfalcou commented on include/boost/simd/arch/common/scalar/function/genmask.hpp:
```diff
@@ -72,39 +72,6 @@ namespace boost { namespace simd { namespace ext
-//                           , bd::scalar_< bs::logical_<A0> >
-//                           , bd::target_< bd::scalar_<bd::unspecified_<Target> > >
-//                           )
-//   {
```
Those version are used in 

include/boost/simd/arch/common/simd/function/insert.hpp

Seems strange this doesnt make some test fail. Care to have a look ?

jtlap replied:
Le 21/08/2016 à 18:46, Joel Falcou a écrit :
>
> In include/boost/simd/arch/common/scalar/function/genmask.hpp 
> <https://github.com/NumScale/boost.simd/pull/229#discussion_r75598393>:
>
> > @@ -72,39 +72,6 @@ namespace boost { namespace simd { namespace ext
> >      }
> >    };
> >
> > -//   // -----------------------------------------------------------------------------------------------
> > -//   // re-targeted genmask for logical
> > -//   BOOST_DISPATCH_OVERLOAD ( genmask_
> > -//                           , (typename A0, typename Target)
> > -//                           , bd::cpu_
> > -//                           , bd::scalar_< bs::logical_<A0> >
> > -//                           , bd::target_< bd::scalar_<bd::unspecified_<Target> > >
> > -//                           )
> > -//   {
>
> Those version are used in
>
> include/boost/simd/arch/common/simd/function/insert.hpp
>
> Seems strange this doesnt make some test fail. Care to have a look ?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub 
> <https://github.com/NumScale/boost.simd/pull/229/files/650e68d8d41b7d7f15f21f47ff8a9479f4c2f2d8..214f5c55902cca1895dba1a66c4e278a942598a4#r75598393>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AAXV7Rf-bZRESZ0Tj7UBL8mZV9YrRUGpks5qiIDVgaJpZM4JpCLY>.
>
the bench of genmask doesnot compile if that can help.
